presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java

```diff
@@ -399,30 +400,30 @@ protected ExtractionResult visitComparisonExpression(ComparisonExpression node,
             Type fieldType = checkedTypeLookup(symbol);
             NullableValue value = normalized.getValue();
 
-            // when the field is BIGINT and the value is DOUBLE, transform the expression in such a way that
-            // the semantics are preserved while doing the comparisons in terms of double
-            // TODO: figure out a way to generalize this for other types
-            if (value.getType().equals(DOUBLE) && fieldType.equals(BIGINT)) {
-                return process(coerceDoubleToLongComparison(normalized), complement);
+            Optional<NullableValue> coercedValue = coerce(value, fieldType);
+            if (coercedValue.isPresent()) {
+                return createComparisonExtractionResult(normalized.getComparisonType(), symbol, fieldType, coercedValue.get().getValue(), complement);
             }
 
-            Optional<NullableValue> coercedValue = coerce(value, fieldType);
-            if (!coercedValue.isPresent()) {
-                return super.visitComparisonExpression(node, complement);
+            Optional<Expression> coercedExpression = coerceComparisonWithRounding(
+                    fieldType, normalized.getNameReference(), value.getType(), value.getValue(), normalized.getComparisonType());
+
+            if (coercedExpression.isPresent()) {
+                return process(coercedExpression.get(), complement);
             }
 
-            return createComparisonExtractionResult(normalized.getComparisonType(), symbol, fieldType, coercedValue.get().getValue(), complement);
+            return super.visitComparisonExpression(node, complement);
         }
 
         private Optional<NullableValue> coerce(NullableValue value, Type targetType)
         {
-            if (!TypeRegistry.canCoerce(value.getType(), targetType)) {
-                return Optional.empty();
-            }
             if (value.isNull()) {
                 return Optional.of(NullableValue.asNull(targetType));
             }
-            Object coercedValue = new FunctionInvoker(metadata.getFunctionRegistry())
+            if (!TypeRegistry.canCoerce(value.getType(), targetType)) {
```

Comment:
I don't think we can move this coercion check after the null check. That would imply that the any null type can be coerced into any other type.